File: plugins/stacks/openclaw/src/shared.ts
import type { TSchema } from 'typebox';
import { formatAccountHistory } from './formatters.js';
import { normalizeSendParams, requireSendParams } from './params.js';
import {
  formatMicroStx,
  getStacksSenderKey,
  getStacksWalletConfig,
  parseStxAmount,
  resolveStacksAddress,
} from './wallet.js';

export type ToolHandler = (params: any) => Promise<any> | any;

export interface ToolSpec {
  name: string;
  label?: string;
  description: string;
  parameters: TSchema;
  handler: ToolHandler;
  /** Mark write tools that require signing (senderKey injected from env). */
  signed?: boolean;
  /** Convert `amount` from STX to microSTX before calling the handler. */
  parseAmount?: boolean;
  /** OpenClaw optional tool registration (write tools). */
  optional?: boolean;
  formatResult?: (result: unknown, params: Record<string, unknown>) => string;
}

export interface ToolRuntimeContext {
  pluginConfig?: { defaultNetwork?: string };
}

function needsAddressResolution(name: string): boolean {
  return (
    name.includes('balance') ||
    name.includes('history') ||
    name.includes('stacking_status') ||
    name === 'stacks_lookup_address' ||
    name === 'stacks_can_stack'
  );
}

function requiresAddress(name: string): boolean {
  return name.includes('balance') || name.includes('history');
}

export function enrichParams(
  spec: ToolSpec,
  params: Record<string, unknown>,
  ctx: ToolRuntimeContext
): Record<string, unknown> {
  const wallet = getStacksWalletConfig(ctx.pluginConfig);
  let enriched =
    spec.name === 'stacks_send_tokens' ? normalizeSendParams(params) : { ...params };

  if (!enriched.network) {
    enriched.network = wallet.network;
  }

  if ('address' in enriched || needsAddressResolution(spec.name)) {
    const resolved = resolveStacksAddress(enriched.address as string | undefined, ctx.pluginConfig);
    if (resolved) {
      enriched.address = resolved;
    } else if (requiresAddress(spec.name)) {
      throw new Error(
        'No wallet address configured. Set STACKS_SENDER_KEY or STACKS_WALLET_ADDRESS in .env.'
      );
    }
  }

  if (spec.signed) {
    const senderKey = getStacksSenderKey();
    if (!enriched.senderKey && senderKey) {
      enriched.senderKey = senderKey;
    }
    if (!enriched.senderKey) {
      throw new Error(
        'No signing key configured. Set STACKS_SENDER_KEY in .env for sends and other write actions.'
      );
    }
  }

  if (spec.parseAmount && enriched.amount != null) {
    const micro = parseStxAmount(enriched.amount as string | number);
    if (!micro) {
      throw new Error(`Could not parse STX amount: ${enriched.amount}`);
    }
    enriched.amount = micro;
  }

  if (spec.name === 'stacks_send_tokens') {
    requireSendParams(enriched);
    if (!spec.parseAmount && enriched.amount != null) {
      const micro = parseStxAmount(enriched.amount as string | number);
      if (micro) enriched.amount = micro;
    }
  }

  return enriched;
}

function defaultFormatResult(spec: ToolSpec, result: unknown): string {
  if (spec.name === 'stacks_get_balance' && result && typeof result === 'object') {
    const b = result as {
      address?: string;
      network?: string;
      stx?: string;
      locked?: string;
      fungibleTokens?: Record<string, { balance: string }>;
    };
    const lines = [
      `Balance for ${b.address} (${b.network})`,
      `Spendable: ${formatMicroStx(b.stx ?? '0')}`,
      `Locked (stacking): ${formatMicroStx(b.locked ?? '0')}`,
    ];
    const tokens = Object.entries(b.fungibleTokens ?? {});
    if (tokens.length > 0) {
      lines.push('', 'Fungible tokens:');
      for (const [id, { balance }] of tokens.slice(0, 5)) {
        lines.push(`  ${id}: ${balance}`);
      }
    }
    return lines.join('\n');
  }

  if (spec.name === 'stacks_send_tokens' && result && typeof result === 'object') {
    const r = result as { success?: boolean; txid?: string; error?: string; reason?: string };
    if (r.success && r.txid) {
      const net = getStacksWalletConfig().network;
      const id = r.txid.startsWith('0x') ? r.txid.slice(2) : r.txid;
      const explorer =
        net === 'mainnet'
          ? `https://explorer.hiro.so/txid/${id}`
          : `https://explorer.hiro.so/txid/${id}?chain=testnet`;
      return `Transfer submitted successfully.\nTxID: ${r.txid}\nExplorer: ${explorer}`;
    }
    return `Transfer failed: ${r.error ?? 'unknown'}${r.reason ? ` (${r.reason})` : ''}`;
  }

  if (spec.name === 'stacks_get_account_history' && result && typeof result === 'object') {
    return formatAccountHistory(result as Parameters<typeof formatAccountHistory>[0]);
  }

  return typeof result === 'string' ? result : JSON.stringify(result, null, 2);
}

export function executeTool(
  spec: ToolSpec,
  rawParams: Record<string, unknown>,
  ctx: ToolRuntimeContext
): Promise<{ text: string; isError: boolean }> {
  return (async () => {
    try {
      const params = enrichParams(spec, rawParams, ctx);
      const result = await spec.handler(params);
      const text = spec.formatResult
        ? spec.formatResult(result, params)
        : defaultFormatResult(spec, result);
      return { text, isError: false };
    } catch (error: any) {
      return {
        text: `Stacks tool ${spec.name} failed: ${error?.message ?? String(error)}`,
        isError: true,
      };
    }
  })();
}

File: plugins/stacks/hermes/shared.py
"""Shared enrichment and execution helpers for Hermes Stacks tools."""

from __future__ import annotations

import json
import os
from typing import Any

from .formatters import format_tool_result
from .params import normalize_send_params, parse_stx_amount, require_send_params
from .wallet import get_network, has_sender_key, resolve_address
from .bridge import run_ts_tool

WRITE_TOOLS = frozenset(
    {
        "stacks_send_tokens",
        "stacks_stack",
        "stacks_delegate_stx",
        "stacks_revoke_delegate",
        "stacks_transfer_name",
        "stacks_contract_call",
        "stacks_swap_execute",
        "stacks_bridge_initiate",
    }
)

PARSE_AMOUNT_TOOLS = frozenset(
    {
        "stacks_send_tokens",
        "stacks_stack",
        "stacks_delegate_stx",
        "stacks_can_stack",
    }
)

OPTIONAL_ADDRESS_TOOLS = frozenset(
    {
        "stacks_get_balance",
        "stacks_get_account_history",
        "stacks_stacking_status",
    }
)


def enrich_params(tool_name: str, params: dict[str, Any]) -> dict[str, Any]:
    enriched = normalize_send_params(params) if tool_name == "stacks_send_tokens" else dict(params)

    if not enriched.get("network"):
        enriched["network"] = get_network()

    if "address" in enriched or tool_name in OPTIONAL_ADDRESS_TOOLS:
        resolved = resolve_address(enriched.get("address"))
        if resolved:
            enriched["address"] = resolved
        elif tool_name in {"stacks_get_balance", "stacks_get_account_history"}:
            raise ValueError(
                "No wallet address configured. Set STACKS_SENDER_KEY or STACKS_WALLET_ADDRESS in .env."
            )

    if tool_name in WRITE_TOOLS:
        if not enriched.get("senderKey") and has_sender_key():
            enriched["senderKey"] = os.environ["STACKS_SENDER_KEY"]
        if not enriched.get("senderKey"):
            raise ValueError(
                "No signing key configured. Set STACKS_SENDER_KEY in .env for sends and other write actions."
            )

    if tool_name in PARSE_AMOUNT_TOOLS and enriched.get("amount") is not None:
        micro = parse_stx_amount(enriched["amount"])
        if not micro:
            raise ValueError(f"Could not parse STX amount: {enriched['amount']}")
        enriched["amount"] = micro

    if tool_name == "stacks_send_tokens":
        require_send_params(enriched)

    return enriched


def run_tool(tool_name: str, params: dict[str, Any]) -> str:
    try:
        enriched = enrich_params(tool_name, params)
        raw = run_ts_tool(tool_name, enriched)
        parsed = json.loads(raw)
        if not parsed.get("success"):
            return json.dumps(parsed)

        result = parsed.get("result")
        text = format_tool_result(tool_name, result, enriched.get("network", get_network()))
        return json.dumps({"success": True, "result": result, "text": text})
    except Exception as exc:  # noqa: BLE001
        return json.dumps({"success": False, "error": str(exc)})
